### PR TITLE
Adding email validation and test method to our business comment and post models.

### PR DESCRIPTION
**Pull request by DamionSilver**
A breakdown of what I did was as follows:

I started by making a branch PostModels . Inside PostModels I created a new branch called FailedTdd. I wrote my initial test and my test failed, because my models class didn't reflect the information needed for the test to pass. I then merged this branch with PostModels. Afterwards, i created a new branch called PassedTdd. I wrote the minimal implementation for my test to pass. I then merged passedtdd with PostModels.

Afterwards, I made a CommentModels branch. Inside CommentModels, I created a new branch called FailedCommentTdd. I wrote my initial test and my test failed, because my models class didn't reflect the information needed for the test to pass. I then merged this branch with CommentModels. Afterwards, I created a new branch called PassedCommentTtd.  I wrote the minimal implementation for my test to pass. I then merged PassedCommentTtd with CommentModels.

Next, I deleted the branches that wasn't in use anymore, which was PassedCommentTtd , FailedCommentTdd,  FailedTdd,  passedtdd.

Lastly, I created a branch called ModelTesting for both of my test (CommentModels and PostModels). I then merged each branch into ModelTesting branch.

### CODE REVIEW

File: Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        public void Comment_Constructor_Pass()
+        {
+            //arrange
+            const string Content = "This is some content";
+            const string userEmail = "damion.silver@test.com";
+            //act
+            Comment newcomment = new Comment(userEmail, Content);
+
+            //assert
+            Assert.NotNull(newcomment);
```
Comment (escalonn) [inline Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs:25]: it's not a problem, but FYI there's no need for any assertion here, since the result of a ctor can't ever be null. leave it out and you're still just testing that the ctor doesn't throw an exception.

File: Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            //arrange
+            const string Content = "This is some content";
+            const string userEmail = "damion.silver@test.com";
+            //act
+            Comment newcomment = new Comment(userEmail, Content);
+
+            //assert
+            Assert.NotNull(newcomment.UserEmail);
+        }
```
Comment (escalonn) [inline Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs:39]: this is probably overdoing it, this test is pretty redundant given the presence of the next test.

File: Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+            Assert.Contains("@", newcomment.UserEmail);
+        }
+
+        [Fact]
+        public void Comment_EmailFormationThrowsException()
+        {
+            const string Content = "This is some content";
+            const string userEmail = "damion.silvertest.com";
+
+            Assert.Throws<ArgumentException>(() => new Comment(userEmail, Content));
```
Comment (escalonn) [inline Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs:86]: prefer ThrowsAny for liskov subsitution principle

File: Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+        {
+            //arrange
+            const string Content = "This is some content";
+            const string userEmail = "damion.silver@test.com";
+            //act
+            Comment newcomment = new Comment(userEmail, Content);
+
+            //assert
+            Assert.Contains("@", newcomment.UserEmail);
+        }
```
Comment (escalonn) [inline Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs:78]: this is redundant if we have the Comment_EmailEquality test.

File: Fakebook.Posts/Fakebook.Posts.Domain/Models/Comment.cs [hunk 5]
```diff
@@ -17,9 +16,17 @@ public Comment(string userEmail, string content)
         {
             if (string.IsNullOrWhiteSpace(userEmail)) throw new ArgumentException("User email is required.", nameof(userEmail));
             if (string.IsNullOrWhiteSpace(content)) throw new ArgumentException("Post content is required.", nameof(content));
+            try
+            {
+                userEmail = new System.Net.Mail.MailAddress(userEmail).Address;
```
Comment (escalonn) [inline Fakebook.Posts/Fakebook.Posts.Domain/Models/Comment.cs:21]: why not have `using System.Net.Mail;` at the top instead of having the full identifier here

**Review by escalonn — CHANGES_REQUESTED** [review]
- in future, request review from me on the mini PRs that built this branch too
- comment test comments apply to the post tests too
- the PR name is bad. it should describe what change is happening to the code.

File: Fakebook.Posts/Fakebook.Posts.IntegrationTests/Properties/launchSettings.json [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    "Fakebook.Posts.IntegrationTests": {
+      "commandName": "Project",
+      "launchBrowser": true,
+      "environmentVariables": {
+        "ASPNETCORE_ENVIRONMENT": "Development"
+      },
+      "applicationUrl": "https://localhost:5001;http://localhost:5000"
+    }
+  }
+}
```
Comment (escalonn) [inline Fakebook.Posts/Fakebook.Posts.IntegrationTests/Properties/launchSettings.json:27]: why is this here

File: Fakebook.Posts/Fakebook.Posts.Domain/Models/Comment.cs [hunk 7]
```diff
@@ -17,9 +17,17 @@ public Comment(string userEmail, string content)
...
+            {
+                throw new ArgumentException("Enter a correct email format.", nameof(userEmail), e);
+            }
             UserEmail = userEmail;
             Content = content;
             Likes = new HashSet<string>();
         }
     }
-}
+}
```
Comment (escalonn) [inline Fakebook.Posts/Fakebook.Posts.Domain/Models/Comment.cs:33]: fix the code smell, don't remove the newline at end of file. same for the post.cs file.

File: Fakebook.Posts/Fakebook.Posts.UnitTests/PostTest.cs [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+        {
+            //arrange
+            const string Content = "This is some content";
+            const string userEmail = "damion.silver@test.com";
+            //act
+            Post newpost = new Post(userEmail, Content);
+        }
+
+        [Fact]
+        public void Post_EmailIsNotNull()
```
Comment (escalonn) [inline Fakebook.Posts/Fakebook.Posts.UnitTests/PostTest.cs:26]: like i said this test isn't doing anything for us, given Post_EmailEquality, so just take it out.
Post_EmailisFormattedCorrectly is also not doing anything Post_EmailEquality is not also testing.

if a method returns "abc", we don't need three unit tests to test each character one by one, one unit test will do.

**Review by escalonn — CHANGES_REQUESTED** [review]
when you push changes that address a previous review comment, click "resolve conversation" on that comment, after leaving a reply indicating how you addressed it if it's not obvious.

the "throws exception" test methods could use more descriptive names or commenting, you have to look at the test for a couple seconds to see what it's about. e.g. renaming the email variable to "invalidEmail" would be an improvement.

File: Fakebook.Posts/Fakebook.Posts.IntegrationTests/Properties/launchSettings.json (hunk 6 above)
Comment (DamionSilver) [reply Fakebook.Posts/Fakebook.Posts.IntegrationTests/Properties/launchSettings.json:27]: I think I clicked my space bar and it detected a change, so when I ran the command git add . this was added to the commit.

File: Fakebook.Posts/Fakebook.Posts.Domain/Models/Comment.cs (hunk 7 above)
Comment (DamionSilver) [reply Fakebook.Posts/Fakebook.Posts.Domain/Models/Comment.cs:33]: I added a new line at the end of this file.

File: Fakebook.Posts/Fakebook.Posts.UnitTests/PostTest.cs (hunk 8 above)
Comment (DamionSilver) [reply Fakebook.Posts/Fakebook.Posts.UnitTests/PostTest.cs:26]: I deleted this method.

File: Fakebook.Posts/Fakebook.Posts.Domain/Models/Comment.cs (hunk 5 above)
Comment (DamionSilver) [reply Fakebook.Posts/Fakebook.Posts.Domain/Models/Comment.cs]: added name space to the top as suggested

File: Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs (hunk 4 above)
Comment (DamionSilver) [reply Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs]: I deleted this method.

File: Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs (hunk 3 above)
Comment (DamionSilver) [reply Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs]: I used ThrowsAny instead of Throw.

File: Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs (hunk 2 above)
Comment (DamionSilver) [reply Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs]: I deleted this method.

File: Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs (hunk 1 above)
Comment (DamionSilver) [reply Fakebook.Posts/Fakebook.Posts.UnitTests/CommentTest.cs]: I erased the assertion.

File: Fakebook.Posts/Fakebook.Posts.IntegrationTests/Properties/launchSettings.json (hunk 6 above)
Comment (DamionSilver) [reply Fakebook.Posts/Fakebook.Posts.IntegrationTests/Properties/launchSettings.json]: I deleted the file from the commit

File: Fakebook.Posts/Fakebook.Posts.UnitTests/PostTest.cs (hunk 8 above)
Comment (escalonn) [reply Fakebook.Posts/Fakebook.Posts.UnitTests/PostTest.cs]: read the comment carefully, i'm also talking about `Post_EmailisFormattedCorrectly`

**Review by escalonn — CHANGES_REQUESTED** [review]
read one of the previous review comments more carefully, i mentioned two tests that don't belong, again

File: Fakebook.Posts/Fakebook.Posts.UnitTests/PostTest.cs (hunk 8 above)
Comment (DamionSilver) [reply Fakebook.Posts/Fakebook.Posts.UnitTests/PostTest.cs]: I removed the formatted email method Post_EmailisFormattedCorrectly and Comment_EmailisFormattedCorrectly method.

File: Fakebook.Posts/Fakebook.Posts.UnitTests/PostTest.cs [hunk 9]
```diff
@@ -22,19 +22,6 @@ public void Post_Constructor_Pass()
...
-            //arrange
-            const string Content = "This is some content";
-            const string userEmail = "damion.silver@test.com";
-            //act
-            Post newpost = new Post(userEmail, Content);
-
-            //assert
-            Assert.Equal(userEmail, newpost.UserEmail);
-        }
-
```
Comment (escalonn) [inline Fakebook.Posts/Fakebook.Posts.UnitTests/PostTest.cs]: why did this one (Post_EmailEquality) get removed, this was a good one. pay attention to the comments, this is very sloppy

**Review by escalonn — CHANGES_REQUESTED** [review]

File: Fakebook.Posts/Fakebook.Posts.UnitTests/PostTest.cs (hunk 9 above)
Comment (DamionSilver) [reply Fakebook.Posts/Fakebook.Posts.UnitTests/PostTest.cs]: I added this test back to the commit. I deleted more than I intended to delete. I am sorry for it being sloppy.

**Review by escalonn — APPROVED** [review]
the name of this PR should be fixed before merging: should mention adding email format validation, and TDD is an unnecessary detail in the title.
